### PR TITLE
feat(internal): template fields enforcement

### PR DESCRIPTION
**Pull request by rarkins**
This PR defines a list of fields that are allowed to be compiled into templates. The idea is partly about security but also about predictability and documentation - I plan to use the exported field names and descriptions to dynamically create a new documentation page for users.

### CODE REVIEW

**Comment by rarkins** [comment]
Design choices / discussions:
- Originally I was planning to do enforcement per-field, e.g. `commitMessage`, `prBanner`, etc. I decided that the difference wasn't enough to justify the complexity.
- I would really like to be able to determine if users have defined template fields that aren't supported, so we can log a warning. This would help me quickly correct any mistakes found/logged in the app too. I'm hoping that Handlebars has a proper way to do this rather than us using hacky regexes.

**Comment by rarkins** [comment]
TODO:
- [ ] Fill in all fields
- [ ] Add code to detects unsupported fields and warn

File: lib/util/template/index.ts [hunk 1]
```diff
@@ -1,7 +1,50 @@
...
+    'The major version of the new version. e.g. "3" if the new version if "3.1.0"',
+  newValue:
+    'The new value in the upgrade. Can be a range or version e.g. "^3.0.0" or "3.1.0"',
+  platform: 'VCS platform in use, e.g. "github", "gitlab", etc.',
+  repository: 'The current repository',
+  toVersion: 'The new version in the upgrade, e.g. "3.1.0"',
+  upgrades: 'An array of upgrade objects in the branch',
+};
+
+function getFilteredObject(obj: any): any {
```
Comment (viceice) [inline lib/util/template/index.ts:28]: I think we should do some type checking here, or atleast at `compile`

**Review by viceice — COMMENTED** [review]
LGTM so far, just need to think how we can do more type safety

File: lib/util/template/index.ts (hunk 1 above)
Comment (rarkins) [reply lib/util/template/index.ts:28]: What would type safety achieve in this case? The fields used in the templates are far more likely to have mistakes, and we can't use type checking for those.

File: lib/util/template/index.ts [hunk 2]
```diff
@@ -1,7 +1,108 @@
+import is from '@sindresorhus/is';
+import safeStringify from 'fast-safe-stringify';
 import * as handlebars from 'handlebars';
+import { logger } from '../../logger';
 
 handlebars.registerHelper('encodeURIComponent', encodeURIComponent);
 
-export function compile(template: string, input: any): string {
+// istanbul ignore next
```
Comment (viceice) [inline lib/util/template/index.ts:8]:
```suggestion
```
This gets calleb by `lib/util/template/index.spec.ts`

File: lib/util/template/index.ts [hunk 3]
```diff
@@ -1,7 +1,108 @@
...
+  references: 'A list of references for the upgrade',
+  releases: 'An array of releases for an upgrade',
+  repository: 'The current repository',
+  toVersion: 'The new version in the upgrade, e.g. "3.1.0"',
+  updateType: 'One of digest, pin, rollback, patch, minor, major',
+  upgrades: 'An array of upgrade objects in the branch',
+};
+
+function getFilteredObject(input: any): any {
+  const obj = JSON.parse(safeStringify(input));
```
Comment (viceice) [inline lib/util/template/index.ts:81]: Use `lib/util/clone` and use `safeStringify` there?

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/util/template/index.ts (hunk 3 above)
Comment (rarkins) [reply lib/util/template/index.ts]: Done

File: lib/util/template/index.ts (hunk 2 above)
Comment (viceice) [reply lib/util/template/index.ts:8]: ?? Shouldn't we have a test for this? To be safe? I know it's trivial but we should use ignore only for untestable scenarios

**Review by viceice — APPROVED** [review]

**Pull request merged**